ch6 in-out-four-normal: add more detail to instructions

satsie opened the pull request:
In this PR I've fleshed out some of the instructions for chapter 6's in-out-four-normal challenge. I'd like to copy some of these updates over to the hard version of the challenge, but will do so as a separate PR.

Before:

![Screenshot from 2024-08-29 14-54-03](https://github.com/user-attachments/assets/fbc0112b-e61b-46a1-897a-7dca7dbd1147)

After:

![Screenshot from 2024-11-06 21-52-19](https://github.com/user-attachments/assets/502c8501-f44f-4b83-93ee-7ef63b73f9ec)

![Screenshot from 2024-11-06 21-56-41](https://github.com/user-attachments/assets/45f270e3-1e4b-4ed7-83c0-d8a3834ac139)

[Preview](https://saving-satoshi-git-fork-satsie-ch6-challen-611092-savingsatoshi.vercel.app/en/chapters/chapter-6/in-out-4-normal?dev=true)

benalleng commented:
Making a note that this relates to one of our checkboxes!
https://github.com/saving-satoshi/saving-satoshi/issues/930#issuecomment-2101034899

benalleng commented:
Do you think we maybe should move some of this to a hint modal? its a lot of text to have on a read-only page

satsie commented:
@benalleng That's fair. I'm happy to move over anything you or anyone else specifically wants over in the hints section. 

Though I feel that precisely because this is a read only exercise, it's not that big of an ask to have the user read up on inputs, outpoints, endianness, and serialization. Nothing new was really added, I just made an effort to clarify what was already there. For example, I think it would be a loss if we didn't at least identify the term endianness since it's something not all developers outside of the Bitcoin industry have to work with. You could go your whole career as a Java dev and never learn what it means.

Is there a worry that if there is too much text the player won't read anything at all? Some additional context is I am writing a detailed curriculum for Art that lists what each chapter covers. I think it would be beneficial for some of these things to be front and center. It's tough to balance education with fun though, and I think any updates I propose to the hard mode of this challenge will not be this detailed (or should they be? should hard mode have more information than easy?). I guess worst case scenario is the player just doesn't read it all, or we can add additional, non coding screens, though that seems like a disproportionately large amount of work for what is gained.

pinheadmz commented on i18n/locales/en.ts:
```diff
@@ -1343,22 +1343,36 @@ const translations = {
-          c: ' when we see them as hexadecimal strings. When we accept hashes as strings from a user we must reverse the byte order before storing or transmitting them as raw bytes. This is why we reverse the byte order of the txid argument that is passed in here.',
+          c: '. This ordering of bytes is called ',
+          d: {
+            text: 'endianness',
```
Only numbers have endianness. It is not correct to say that a hash is big or little endian. Bitcoin block hashes actually come out of the hash function with the 0's on the *right side*. Satoshi decided to interpret that as a little-endian integer. Sometime later when the JSON-RPC was eventually added, the decision was made to display those block hashes backwards so that they looked more like conventional big endian integers.

So what we do to hashes (sipa would say) is "byte swap" or just "reverse" them.

satsie replied:
Endianness only applies to numbers. Endianness only applies to numbers. Endianness only applies to numbers. 

I need to get a tattoo of this!!!!! Ty for the feedback! To refresh my memory, I also dug up some old comments you had  made that basically reiterate what you said here. 

I'll update this PR to use "byte swap" or "reverse"

pinheadmz reviewed (commented):
I'm not sure about the endianness changes (see comment) but I do like a deeper explanation of outpoint. Maybe the endianness / reverse-hash section can be reworked as a pop up hint or on the hints page

satsie commented:
Updated this PR to remove the incorrect application of "endianness" to hashes. Also added a ChatBTC tool tip for endianness at the end where we remind the user that integers in bitcoin are little-endian.

There is no existing help page for this screen because it's not an interactive exercise. If we decide to move anything out to a help page, can that be done in a separate PR?

pinheadmz commented on i18n/locales/en.ts:
```diff
@@ -1358,22 +1358,30 @@ const translations = {
         heading: 'Looking at the Input class implementation',
         paragraph_one:
-          'Bitcoin transaction inputs always point to existing, unspent transaction outputs. Therefore, our Input class has a method <span className="text-green"> from_output() </span> which is used to construct an Input by passing in the output description:',
+          'Here we have code for two classes: an <span className="font-bold">Input</span> class and an <span className="font-bold">Outpoint</span> class.',
```
Maybe something like this so right at the top, the distinction is clear. Otherwise its so easy to mis-read 

```suggestion
          'Here we have code for two classes: an <span className="font-bold">Input</span> class and an <span className="font-bold">Outpoint</span> (not "output"!) class.',
```

satsie replied:
Ah good point!!! It's funny how normal these small distinctions become to us after a while :)

pinheadmz commented on i18n/locales/en.ts:
```diff
@@ -1358,22 +1358,30 @@ const translations = {
             href: 'https://chat.bitcoinsearch.xyz/?author=holocat&question=why%2520are%2520hashes%2520reversed%2520in%2520bitcoin',
           },
-          c: ' when we see them as hexadecimal strings. When we accept hashes as strings from a user we must reverse the byte order before storing or transmitting them as raw bytes. This is why we reverse the byte order of the txid argument that is passed in here.',
+          c: `. When a hash is provided in hexadecimal format, the byte order must be reversed before storing or transmitting the data as raw bytes.`,
```
```suggestion
          c: `, but only when presented to or entered by a user. When a hash is provided in hexadecimal format, the byte order must be reversed before storing or transmitting the data as raw bytes.`,
```

satsie replied:
Thanks for catching this!

pinheadmz reviewed (approved):
Great! thanks.

satsie commented on i18n/locales/en.ts:
```diff
@@ -1358,22 +1358,30 @@ const translations = {
+        paragraph_six:
+          'Together, these two pieces of information make up an <span className="font-bold">Outpoint</span>. Eventually we will pass in the txid and vout values that came from executing the <span className="text-green p-1 font-mono bg-[#00000033] m-1 text-base">listunspent</span> command in the previous exercise. ',
+        paragraph_seven:
+          "The second two arguments are the value of the output we want to spend (in satoshis) and something called a scriptcode. That data is not needed until later so let's temporarily store it as an empty byte array.",
```
```suggestion
          "The second two arguments are the value of the output we want to spend (in satoshis) and something called a scriptcode. That data is not needed until later so let's temporarily use an empty byte array.",
```
